File: sdk/typescript/test/native.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { resolve, join } from 'node:path';
import { mkdtemp, writeFile, readFile, mkdir } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';
import { NativeBackend } from '../src/native.js';

const FIXTURES = resolve(import.meta.dirname, '../../contract-tests/fixtures');

describe('NativeBackend', () => {
  describe('simple-store', () => {
    const backend = new NativeBackend(resolve(FIXTURES, 'simple-store'));

    it('reads a string value', async () => {
      expect(await backend.get('cluster', 'name')).toBe('noizu');
    });

    it('reads a nested string', async () => {
      expect(await backend.get('cluster', 'node_pool.instance_type')).toBe('m5.xlarge');
    });

    it('reads an integer', async () => {
      expect(await backend.get('cluster', 'port')).toBe(6443);
    });

    it('reads a boolean', async () => {
      expect(await backend.get('cluster', 'enabled')).toBe(true);
    });

    it('reads entire config as object', async () => {
      const result = await backend.get('cluster');
      expect(result).toBeTypeOf('object');
      expect(result).toHaveProperty('name');
      expect(result).toHaveProperty('node_pool');
    });

    it('returns null for missing path', async () => {
      expect(await backend.get('cluster', 'nonexistent')).toBeNull();
    });

    it('lists configs', async () => {
      const configs = await backend.listConfigs();
      expect(configs).toEqual(['cluster']);
    });
  });

  describe('nested-store', () => {
    const backend = new NativeBackend(resolve(FIXTURES, 'nested-store'));

    it('reads array element by index', async () => {
      expect(await backend.get('app', 'endpoints[0].host')).toBe('api.example.com');
    });

    it('reads with negative index', async () => {
      expect(await backend.get('app', 'endpoints[-1].host')).toBe('backup.example.com');
    });

    it('reads with wildcard', async () => {
      expect(await backend.get('app', 'endpoints[*].host')).toEqual([
        'api.example.com',
        'internal.example.com',
        'backup.example.com',
      ]);
    });

    it('reads length', async () => {
      expect(await backend.get('app', 'endpoints.length')).toBe(3);
    });

    it('reads chained brackets', async () => {
      expect(await backend.get('app', 'matrix[0][1]')).toBe(2);
    });

    it('lists configs', async () => {
      const configs = await backend.listConfigs();
      expect(configs).toEqual(['app']);
    });
  });

  describe('write operations', () => {
    let store: string;
    let backend: NativeBackend;

    beforeEach(async () => {
      store = await mkdtemp(join(tmpdir(), 'dc-native-write-'));
      await mkdir(join(store, 'test'), { recursive: true });
      await writeFile(join(store, 'test', 'base.yaml'), stringifyYaml({ existing: 'value' }));
      await writeFile(join(store, 'test', '.active'), stringifyYaml({ existing: 'value' }));
      await writeFile(join(store, '.meta'), stringifyYaml({ configs: ['test'] }));
      backend = new NativeBackend(store);
    });

    it('set writes to layer and updates .active', async () => {
      await backend.set('test', 'newkey', 'hello');

      const layerContent = await readFile(join(store, 'test', 'local.yaml'), 'utf-8');
      const layer = parseYaml(layerContent);
      expect(layer).toHaveProperty('newkey', 'hello');

      const activeContent = await readFile(join(store, 'test', '.active'), 'utf-8');
      const active = parseYaml(activeContent);
      expect(active).toHaveProperty('existing', 'value');
      expect(active).toHaveProperty('newkey', 'hello');
    });

    it('set with noBump does not change version', async () => {
      await backend.set('test', 'key', 'val', 'local', true);

      // version file should not exist since we started fresh and used noBump
      const { readVersion } = await import('../src/version.js');
      const version = await readVersion(store);
      expect(version).toBe(0);
    });

    it('set with custom layer writes to that layer file', async () => {
      await backend.set('test', 'secret', 'password', 'secrets');

      const layerContent = await readFile(join(store, 'test', 'secrets.yaml'), 'utf-8');
      const layer = parseYaml(layerContent);
      expect(layer).toHaveProperty('secret', 'password');
    });

    it('unset removes key and updates .active', async () => {
      // First set a key in local layer
      await backend.set('test', 'toremove', 'gone', 'local', true);

      // Now unset it
      await backend.unset('test', ['toremove'], 'local', true);

      const layerContent = await readFile(join(store, 'test', 'local.yaml'), 'utf-8');
      const layer = parseYaml(layerContent);
      expect(layer).not.toHaveProperty('toremove');
    });

    it('unset on missing layer is no-op', async () => {
      // Should not throw when layer file does not exist
      await expect(backend.unset('test', ['anykey'], 'nonexistent')).resolves.toBeUndefined();
    });

    it('bump increments version', async () => {
      const v1 = await backend.bump();
      expect(v1).toBe(1);
      const v2 = await backend.bump();
      expect(v2).toBe(2);
    });
  });
});

File: sdk/elixir/test/direnv_config/native_test.exs
defmodule DirenvConfig.NativeTest do
  use ExUnit.Case, async: false

  @fixtures_dir Path.expand("../../../contract-tests/fixtures", __DIR__)

  describe "get/3 with simple-store" do
    @store Path.join(@fixtures_dir, "simple-store")

    test "reads a simple string" do
      assert DirenvConfig.Native.get(@store, "cluster", "name") == {:ok, "noizu"}
    end

    test "reads nested string" do
      assert DirenvConfig.Native.get(@store, "cluster", "node_pool.instance_type") ==
               {:ok, "m5.xlarge"}
    end

    test "reads integer" do
      assert DirenvConfig.Native.get(@store, "cluster", "port") == {:ok, 6443}
    end

    test "reads boolean" do
      assert DirenvConfig.Native.get(@store, "cluster", "enabled") == {:ok, true}
    end

    test "reads entire config as map" do
      assert {:ok, data} = DirenvConfig.Native.get(@store, "cluster")
      assert is_map(data)
      assert Map.has_key?(data, "name")
      assert Map.has_key?(data, "node_pool")
    end

    test "missing key returns error" do
      assert DirenvConfig.Native.get(@store, "cluster", "nonexistent") == :error
    end
  end

  describe "get/3 with nested-store" do
    @nested_store Path.join(@fixtures_dir, "nested-store")

    test "array index" do
      assert DirenvConfig.Native.get(@nested_store, "app", "endpoints[0].host") ==
               {:ok, "api.example.com"}
    end

    test "wildcard" do
      assert {:ok, hosts} =
               DirenvConfig.Native.get(@nested_store, "app", "endpoints[*].host")

      assert hosts == ["api.example.com", "internal.example.com", "backup.example.com"]
    end

    test "length" do
      assert DirenvConfig.Native.get(@nested_store, "app", "endpoints.length") == {:ok, 3}
    end

    test "chained brackets" do
      assert DirenvConfig.Native.get(@nested_store, "app", "matrix[0][1]") == {:ok, 2}
    end
  end

  describe "list_configs/1" do
    test "simple store" do
      assert DirenvConfig.Native.list_configs(Path.join(@fixtures_dir, "simple-store")) ==
               {:ok, ["cluster"]}
    end

    test "nested store" do
      assert DirenvConfig.Native.list_configs(Path.join(@fixtures_dir, "nested-store")) ==
               {:ok, ["app"]}
    end
  end

  describe "set/6" do
    setup do
      tmp = Path.join(System.tmp_dir!(), "dc_native_test_#{System.unique_integer([:positive])}")
      File.mkdir_p!(tmp)

      on_exit(fn ->
        File.rm_rf!(tmp)
      end)

      {:ok, store: tmp}
    end

    test "set writes to layer and updates .active", %{store: store} do
      assert :ok = DirenvConfig.Native.set(store, "myconfig", "db.host", "pg.local")

      # Verify the local layer was written
      layer_file = Path.join([store, "myconfig", "local.yaml"])
      assert File.exists?(layer_file)
      {:ok, layer_data} = YamlElixir.read_from_file(layer_file)
      assert layer_data["db"]["host"] == "pg.local"

      # Verify .active was created
      active_file = Path.join([store, "myconfig", ".active"])
      assert File.exists?(active_file)

      # Verify version was bumped
      assert DirenvConfig.Version.read(store) == 1
    end

    test "set with no_bump skips version bump", %{store: store} do
      assert :ok = DirenvConfig.Native.set(store, "myconfig", "key", "val", "local", true)
      assert DirenvConfig.Version.read(store) == 0
    end

    test "set to a specific layer", %{store: store} do
      assert :ok = DirenvConfig.Native.set(store, "myconfig", "env", "production", "base")

      layer_file = Path.join([store, "myconfig", "base.yaml"])
      {:ok, data} = YamlElixir.read_from_file(layer_file)
      assert data["env"] == "production"
    end
  end

  describe "unset/5" do
    setup do
      tmp = Path.join(System.tmp_dir!(), "dc_native_unset_#{System.unique_integer([:positive])}")
      File.mkdir_p!(tmp)

      on_exit(fn ->
        File.rm_rf!(tmp)
      end)

      {:ok, store: tmp}
    end

    test "unset removes key and updates .active", %{store: store} do
      # First set some keys
      :ok = DirenvConfig.Native.set(store, "cfg", "a", "1")
      :ok = DirenvConfig.Native.set(store, "cfg", "b", "2")

      version_before = DirenvConfig.Version.read(store)

      # Unset one key
      assert :ok = DirenvConfig.Native.unset(store, "cfg", ["a"])

      # Verify key was removed from layer
      layer_file = Path.join([store, "cfg", "local.yaml"])
      {:ok, data} = YamlElixir.read_from_file(layer_file)
      refute Map.has_key?(data, "a")
      assert data["b"] == 2

      # Verify version was bumped
      assert DirenvConfig.Version.read(store) > version_before
    end

    test "unset with no_bump skips version bump", %{store: store} do
      :ok = DirenvConfig.Native.set(store, "cfg", "x", "1")
      version_before = DirenvConfig.Version.read(store)

      assert :ok = DirenvConfig.Native.unset(store, "cfg", ["x"], "local", true)
      assert DirenvConfig.Version.read(store) == version_before
    end

    test "unset on nonexistent layer file is ok", %{store: store} do
      assert :ok = DirenvConfig.Native.unset(store, "cfg", ["x"])
    end
  end

  describe "bump/1" do
    setup do
      tmp = Path.join(System.tmp_dir!(), "dc_native_bump_#{System.unique_integer([:positive])}")
      File.mkdir_p!(tmp)

      on_exit(fn ->
        File.rm_rf!(tmp)
      end)

      {:ok, store: tmp}
    end

    test "bump increments version", %{store: store} do
      assert {:ok, 1} = DirenvConfig.Native.bump(store)
      assert {:ok, 2} = DirenvConfig.Native.bump(store)
    end
  end
end
